p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,483 @@
+	}
+	cert, err := webrtc.GenerateCertificate(pk)
+	if err != nil {
+		return nil, fmt.Errorf("generate certificate: %w", err)
+	}
+	config := webrtc.Configuration{
+		Certificates: []webrtc.Certificate{*cert},
+	}
+	noiseTpt, err := noise.New(noise.ID, privKey, nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create noise transport: %w", err)
+	}
+	transport := &WebRTCTransport{
+		rcmgr:        rcmgr,
+		webrtcConfig: config,
+		privKey:      privKey,
+		noiseTpt:     noiseTpt,
+		localPeerId:  localPeerId,
+
+		peerConnectionTimeouts: IceTimeouts{
+			Disconnect: DefaultDisconnectedTimeout,
+			Failed:     DefaultFailedTimeout,
+			Keepalive:  DefaultKeepaliveTimeout,
+		},
+
+		maxInFlightConnections: DefaultMaxInFlightConnections,
+	}
+	for _, opt := range opts {
+		if err := opt(transport); err != nil {
+			return nil, err
+		}
+	}
+	return transport, nil
+}
+
+func (t *WebRTCTransport) Protocols() []int {
+	return []int{ma.P_WEBRTC}
+}
+
+func (t *WebRTCTransport) Proxy() bool {
+	return false
+}
+
+func (t *WebRTCTransport) CanDial(addr ma.Multiaddr) bool {
+	return dialMatcher.Matches(addr)
+}
+
+func (t *WebRTCTransport) Listen(addr ma.Multiaddr) (tpt.Listener, error) {
+	addr, wrtcComponent := ma.SplitLast(addr)
+	isWebrtc := wrtcComponent.Equal(ma.StringCast("/webrtc"))
+	if !isWebrtc {
+		return nil, fmt.Errorf("must listen on webrtc multiaddr")
+	}
+	nw, host, err := manet.DialArgs(addr)
+	if err != nil {
+		return nil, fmt.Errorf("listener could not fetch dialargs: %w", err)
+	}
+	udpAddr, err := net.ResolveUDPAddr(nw, host)
+	if err != nil {
+		return nil, fmt.Errorf("listener could not resolve udp address: %w", err)
+	}
+
+	socket, err := net.ListenUDP(nw, udpAddr)
+	if err != nil {
+		return nil, fmt.Errorf("listen on udp: %w", err)
+	}
+
+	// construct multiaddr
+	listenerMultiaddr, err := manet.FromNetAddr(socket.LocalAddr())
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+
+	listenerFingerprint, err := t.getCertificateFingerprint()
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+
+	encodedLocalFingerprint, err := internal.EncodeDTLSFingerprint(listenerFingerprint)
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+
+	certMultiaddress, err := ma.NewMultiaddr(fmt.Sprintf("/webrtc/certhash/%s", encodedLocalFingerprint))
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+
+	listenerMultiaddr = listenerMultiaddr.Encapsulate(certMultiaddress)
+
+	listener, err := newListener(
+		t,
+		listenerMultiaddr,
+		socket,
+		t.webrtcConfig,
+	)
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+	return listener, nil
+}
+
+func (t *WebRTCTransport) Dial(ctx context.Context, remoteMultiaddr ma.Multiaddr, p peer.ID) (tpt.CapableConn, error) {
+	scope, err := t.rcmgr.OpenConnection(network.DirOutbound, false, remoteMultiaddr)
+	if err != nil {
+		return nil, err
+	}
+	err = scope.SetPeer(p)
+	if err != nil {
+		return nil, err
```

Comment:
We need to close the scope here. See my other comment about splitting this up so
that we can't forget to clean up resources in the case of an error.